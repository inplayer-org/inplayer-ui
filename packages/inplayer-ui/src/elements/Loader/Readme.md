A simple spinner/loader to use as fallback when doing async operations.
By default it inherits the `primary.main` theme color.
Can be passed a `color` prop with a custom hex color.

```jsx static
import { Spinner } from '@inplayer-org/inplayer-ui';

function Page({ isLoading }) {
  if (isLoading) {
    return <Spinner />;
  }

  return (/* Your page content */);
}
```

### Demo

```jsx
<Typography variant="h6">Standard loader</Typography>
<Loader />

<Typography variant="h6">Loader with custom color</Typography>
<Loader color="#8cb" />
```
