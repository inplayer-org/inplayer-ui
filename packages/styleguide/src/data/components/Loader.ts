import { ComponentData } from "../types";

const usage = `
  import { Spinner } from '@inplayer-org/inplayer-ui';

  function Page({ isLoading }) {
    if (isLoading) {
      return <Spinner />;
    }

    return (/* Your page content */);
  }
`;

const demoCode = `
<div>
  <Typography variant="h6">Standard loader</Typography>
  <Loader />

  <Typography variant="h6">Loader with custom color</Typography>
  <Loader color="#8cb" />
</div>
`;

const Loader: ComponentData = {
  description: `A simple spinner/loader to use as fallback when doing async operations. By default it inherits the primary.main theme color. Can be passed a color prop with a custom hex color.`,
  path: `src/elements/Loader/index.js`,
  usage,
  examples: [
    {
      title: 'Demo',
      code: demoCode,
    }
  ],
}

export default Loader;
