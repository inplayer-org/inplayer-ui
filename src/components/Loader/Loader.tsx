import React from 'react';
import styled, { keyframes } from 'styled-components';
import colors from '../../theme/colors';

const directions: Record<string, any> = {
  up: '16,0 32,32 0,32',
  down: '0,0 32,0 16,32',
  right: '0,0 32,16 0,32',
  left: '0,16 32,0 32,32',
};

const dash = keyframes`
  to {
      stroke-dashoffset: 136;
  }
`;

const StyledSpinner = styled.svg<{
  color?: string;
}>`
  transform-origin: 50% 65%;
  stroke: ${({ color }) => color};

  polygon {
    stroke-dasharray: 17;
    animation: ${dash} 2.5s cubic-bezier(0.35, 0.04, 0.63, 0.95) infinite;
  }
`;

type Props = {
  height?: number;
  width?: number;
  color?: string;
  lineWidth?: number;
  direction?: string;
};

const Loader = ({
  height = 100,
  width = 100,
  color = `${colors.skyBlue}`,
  lineWidth = 2,
  direction = 'right',
}: Props) => (
  <StyledSpinner color={color} id="triangle" width={width} height={height} viewBox="-3 -4 39 39">
    <polygon
      fill="transparent"
      stroke={color}
      strokeWidth={lineWidth}
      points={directions[direction]}
    />
  </StyledSpinner>
);

export default Loader;
