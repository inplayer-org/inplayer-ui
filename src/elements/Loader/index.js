// @flow
import React from 'react';
import styled, { keyframes } from 'styled-components';

interface Props {
  height?: number;
  width?: number;
  color?: string;
  lineWidth?: number;
  direction?: string;
}

const directions = {
  up: '16,0 32,32 0,32',
  down: '0,0 32,0 16,32',
  right: '0,0 32,16 0,32',
  left: '0,16 32,0 32,32',
};

const dash = keyframes`
    to {
        stroke-dashoffset: 136;
    }
`;

const StyledSpinner = styled.svg`
  transform-origin: 50% 65%;

  polygon {
    stroke-dasharray: 17;
    animation: ${dash} 2.5s cubic-bezier(0.35, 0.04, 0.63, 0.95) infinite;
  }
`;

const Loader = ({
  height = 100,
  width = 100,
  color = '#07AAE6',
  lineWidth = 2,
  direction = 'right',
}: Props) => (
  <StyledSpinner id="triangle" width={width} height={height} viewBox="-3 -4 39 39">
    <polygon
      fill="transparent"
      stroke={color}
      strokeWidth={lineWidth}
      points={directions[direction]}
    />
  </StyledSpinner>
);

export default Loader;
