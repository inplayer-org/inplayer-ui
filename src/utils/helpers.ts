import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
