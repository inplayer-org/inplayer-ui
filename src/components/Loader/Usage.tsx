import React from 'react';
import Loader from './Loader';

const App = () => <Loader color="red" lineWidth={5} direction="up" height={150} width={150} />;

export default App;
