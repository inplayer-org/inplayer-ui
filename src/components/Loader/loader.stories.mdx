import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Loader from './'

<Meta title="Loder" component={<Loader/>} />

# Loader

A simple spinner/loader to use as fallback when doing async operations. By default it inherits the `primary.main` theme color. 

Can be passed a `color` prop with a custom hex color.

```jsx
import { Loader } from '@inplayer-org/inplayer-ui';

const Page = () => (
  <Loader />
)
```

### Props
<Props of={Loader} />

### Default Loader
<Preview withSource="open">
  <Story name="default">
    <Loader />
  </Story>
</Preview>

### Modified Loader
<Preview withSource="open">
  <Story name="modified">
    <Loader color="black" lineWidth={5} direction="up" height={150} width={150} />
  </Story>
</Preview>
