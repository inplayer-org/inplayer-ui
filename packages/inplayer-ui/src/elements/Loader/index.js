import styled, { keyframes } from 'styled-components';
import { prop, ifProp } from 'styled-tools';
import { uiColors } from 'utils';
import { transparentize } from 'polished';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

const Loader = styled.div`
  border-radius: 50%;
  width: 6em;
  height: 6em;
  font-size: 10px;
  border: 1.1em solid
    ${({ theme, color }) => transparentize(0.8, color || theme.palette.primary.main)};
  border-left: 1.1em solid ${ifProp('color', prop('color'), uiColors('primary.main'))};
  animation: ${spin} 0.8s infinite linear;

  &::after {
    width: 8em;
    height: 8em;
  }
`;

export default Loader;
