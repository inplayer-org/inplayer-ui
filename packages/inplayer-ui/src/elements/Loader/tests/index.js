import { generateSnapshots } from 'packages/inplayer-ui/src/elements/Loader/tests/helpers';

import Loader from '../index';

generateSnapshots(Loader);
