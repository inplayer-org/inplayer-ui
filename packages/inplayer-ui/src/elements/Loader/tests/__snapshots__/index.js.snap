// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader renders correctly 1`] = `
.c0 {
  border-radius: 50%;
  width: 6em;
  height: 6em;
  font-size: 10px;
  border: 1.1em solid rgba(7,170,230,0.2);
  border-left: 1.1em solid #07AAE6;
  -webkit-animation: iECmZH 0.8s infinite linear;
  animation: iECmZH 0.8s infinite linear;
}

.c0::after {
  width: 8em;
  height: 8em;
}

<div
  className="c0"
/>
`;
